doc/stream-telemetry/stream-telemetry-hld.md

```diff
@@ -0,0 +1,986 @@
+    end
+
+    subgraph SONiC service
+    subgraph GNMI container
+        gnmi(GNMI server)
+        counter_syncd(Counter Syncd)
+    end
+    subgraph SWSS container
+    subgraph Orchagent
+        st_orch(Stream Telemetry Orch)
+    end
+    end
+    subgraph SYNCD container
+        syncd(Syncd)
+    end
+    end
+
+    subgraph Linux Kernel
+        dma_engine(DMA Engine)
+        netlink_module(Netlink Module)
+    end
+
+    asic[\ASIC\]
+
+    config_db --STREAM_TELEMETRY_PROFILE
+                STREAM_TELEMETRY_GROUP--> st_orch
+    st_orch --SAI_OBJECT_TYPE_TAM_REPORT
+                SAI_OBJECT_TYPE_TAM_TEL_TYPE
+                SAI_OBJECT_TYPE_TAM_TRANSPORT
+                SAI_OBJECT_TYPE_TAM_COLLECTOR
+                SAI_OBJECT_TYPE_TAM_EVENT
+                SAI_OBJECT_TYPE_TAM--> syncd
+    syncd --IPFIX template--> st_orch
+    syncd --TAM configuration--> dma_engine
+    syncd --TAM configuration--> netlink_module
+    st_orch --STREAM_TELEMETRY_SESSION--> state_db
+    state_db --STREAM_TELEMETRY_SESSION--> counter_syncd
+    asic --counters--> dma_engine
+    dma_engine --IPFIX record--> netlink_module
+    netlink_module --IPFIX record--> counter_syncd
+    counter_syncd --telemetry message--> gnmi
+```
+
+STATE_DB channel model? Produce Table/Consume Table
+
+## High-Level Design
+
+### Modules
+
+#### Counter Syncd
+
+The `counter syncd` is a new module that runs within the GNMI container. Its primary responsibility is to receive counter messages via netlink and convert them into GNMI messages for an external collector. It subscribes to a socket of a specific family and multicast group of generic netlink. The configuration for generic netlink is defined as constants in `/etc/sonic/constants.yml` as follows.
+
+``` yaml
+constants:
+    stream_telemetry:
+        genl_family: "sonic_stel"
+        genl_multicast_group: "stats"
+}
+
+```
+
+#### Stream Telemetry Orch
+
+The `Stream Telemetry Orch` is a new object within the Orchagent. It has following primary duties:
+
+1. Maintain the TAM SAI objects according to the stream telemetry configuration in the config DB.
+2. Generate a unique template ID for each stream telemetry profile to ensure distinct identification and management.
+3. Register and activate streams on counter syncd.
+
+`Stream Telemetry Orch` leverages `tam_counter_subscription` objects to bind monitoring objects, such as ports, buffers, or queues, to streams. Therefore, this orch must ensure that the lifecycle of `tam_counter_subscription` objects is within the lifecycle of their respective monitoring objects.
+
+#### Netlink Module and DMA Engine
+
+These two modules need to be provided by vendors. This document proposes a ring buffer communication model to support all expected TAM configurations as follows.
+
+![netlink_dma_channel](netlink_dma_channel.drawio.svg)
+
+### Data format
+
+We will use IPFIX as the report format, with all numbers in the IPFIX message in network-order (Big-endian).
```

Comment:
You cannot make the assumption that data will be Big- or Little-endian. If it is collected by the ASIC and DMAd to a ring buffer, then kernel driver would need  to correct the endianness of each data set. We need to have a bit in the element ID or Enterprise number that would indicate endianness.

Reply:
Are you worried that endian conversion will consume too many CPU clocks?